chore: Add OpenAPI naming ADR (#1159)
SAP/cloud-sdk-js: knowledge-base/adr/0019-openapi-generator-naming.md

### Issues with properties and parameters

Given the following two types and their original properties, if those were used in combination, e. g. `allOf`, there is no possibility to set both values.

```ts
type A {
  someProperty: string; // original name: 'some_property`
}

type A {
  someProperty: string; // original name: 'some-property`
}
```

This could be solved by making names unique throughout the whole service.
However this again would lead to less convenient names, e.g. `name`, `name1`, `name2`.

## Decisions

### API naming

APIs are named based on the first tag, if there are any.
If there is no tag, a "default" tag is used.
To allow customizing of APIs, we introduced the `x-sap-cloud-sdk-api-name` extension, which takes precedence over the default name generation.
This extension can be set on on multiple levels:

- operation, causing this operation to be treated as part of the given API.
- path, causing all operations under this path to be treated as part of the given API.
- document root, causing all operations in the document to be treated as part of the given API.

All names are transformed to pascal case (`SomeApi`).

> API names are formed by removing "api" suffix (case independant and only if available) and adding "Api".

Author: ```suggestion
API names are formed by appending an "Api" suffix.
If the original name ends with an "api" suffix (case independent), the original suffix is removed prior to appending "Api".
```

Examples:

- `my-api` => `MyApi`
- `my` => `MyApi`

> - `my-api-api` => `MyApiApi`

Reviewer: this contradicts the: 
> by` removing "api" suffix (case independant and only if available)

Author: But also 
> and adding "Api"

I improved the text a little, I hope it is more on point now?

### Operation naming

Operations are named based on their given `operationId`.

> If no `operationId` is given the name is parsed from the method and the path under which the operation occured.

Reviewer: Could you perhaps given an example here of path, method etc and what will be the name generated from that.

Author: ```suggestion
If no `operationId` is given the name is parsed from the method and the path under which the operation occured.
Example: 

- path: `entity/{id}`
- method: `get`
- resulting name: `getEntityById`
```

Users can set specific operation names using the `x-sap-cloud-sdk-operation-name` extension.
This extension can be set on the operation only.

> Names will be transformed to camel case, duplicate names will cause generation to fail.

Reviewer: I wonder if the reformatting to camel case for the API and Operation is good? What are the values for the `operationId` and `tag` in general? Would perhaps be more consistent to also keep the names?

Reviewer: Also mention the `strictNames` generator flag somewhere in the document.

Author: As this is just the ADR, I would like to leave those kinds of details for the documentation portal. Would that be ok for you?

Author: > I wonder if the reformatting to camel case for the API and Operation is good? What are the values for the `operationId` and `tag` in general? Would perhaps be more consistent to also keep the names?

For APIs we have to do this, as tags can be anything, e. g. also containing spaces and APIs are technically just variables and they have to have a valid name. Also, there is nothing comparable to operationId, that would allow extracting technical names for our purposes. Is is a little different with the extensions. Here we could consider using them as is and throwing an error if they are incorrect. 
For operations it is a little different, those could technically keep their names (even if there are spaces) and the `operationId`s are intended to be used for technical names.
However, this might lead to APIs that on the first glance look a little weird, e.g.:

```ts
const SomeApi = {
  a: () => 1,
  'a-2': () => 2
};

SomeApi.a();

// meh
SomeApi['a-2']();
```

The second invocation here is a little unusual. This totally works, but for me personally this is a bit awkward for functions.
The same can occur when accessing properties in objects (e.g. request bodies):

```ts
const body = {
  b: 1,
  'b-2': 2
}

const someVal1 = body.b; 
const someVal2 = body['b-2']; 
```

But in my personal opinion this is less awkward.

Still, this is a valid question, the best way to decide this would be to ask users...

### Property and query parameter naming

Keep the names as provided per the specification.

> No camel case renaming is done.

Reviewer: ```suggestion
No camel case renaming is done to avoid the original name corner cases discussed above.
```

## Consequences

> 

Author: I added the second sentence, but not the first one as this is the `Consequences` section and the sentence speaks about the `Decision` secsion.
